src/main/java/org/testng/TestNG.java

```diff
@@ -275,23 +275,28 @@ public void initializeSuitesAndJarFile() {
 
     m_isInitialized = true;
     if (m_suites.size() > 0) {
-    	//to parse the suite files (<suite-file>), if any
-    	for (XmlSuite s: m_suites) {
-        for (String suiteFile : s.getSuiteFiles()) {
-            Path rootPath = Paths.get(s.getFileName()).getParent();
-            try {
-                Collection<XmlSuite> childSuites = getParser(rootPath.resolve(suiteFile).normalize().toString()).parse();
-                for (XmlSuite cSuite : childSuites){
-                    cSuite.setParentSuite(s);
-                    s.getChildSuites().add(cSuite);
-                }
-            } catch (ParserConfigurationException | IOException | SAXException e) {
-                e.printStackTrace(System.out);
-            }
-        }
-
-    	}
-      return;
+	    //to parse the suite files (<suite-file>), if any
+		for (XmlSuite s: m_suites) {
+		   for (String suiteFile : s.getSuiteFiles()) {
+	        	Path suitePath = Paths.get(s.getFileName());
+	        	Path rootPath = suitePath.getParent();
+	        	try {
+	        		Collection<XmlSuite> childSuites = getParser(rootPath.resolve(suiteFile).normalize().toString()).parse();
+	        		for (XmlSuite cSuite : childSuites){
+	        			Path childSuite = Paths.get(cSuite.getFileName()).normalize();
+	        			Path parentSuite = Paths.get(suiteFile).normalize();
+	        			if(!childSuite.getFileName().equals(parentSuite.getFileName()))
```

Comment:
> However, this is only one level of nesting, what about multiple level of nesting?

For sure, we need some recursive here.

> I still think that if Parser class is not meant to be used by everyone then is this much change worth?

I'm agree. But we can add a new method, or an util method that will filter the existing one.
I think the parsing logic should be located in the same class.